src/WebJobs.Script.WebHost/Diagnostics/DeferredLoggerProvider.cs

```diff
@@ -0,0 +1,112 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+
+namespace Microsoft.Azure.WebJobs.Script.WebHost.Diagnostics
+{
+    public class DeferredLoggerProvider : ILoggerProvider
+    {
+        private readonly Channel<DeferredLogEntry> _channel = Channel.CreateBounded<DeferredLogEntry>(new BoundedChannelOptions(150)
+        {
+            FullMode = BoundedChannelFullMode.DropOldest,
+            // Avoids locks and interlocked operations when reading from the channel.
+            SingleReader = true,
+            SingleWriter = false
+        });
+
+        private bool _isEnabled = true;
+        private bool _disposed = false;
+
+        public int Count => _channel.Reader.Count;
+
+        public ILogger CreateLogger(string categoryName)
+        {
+            if (_isEnabled)
+            {
+                return new DeferredLogger(_channel, categoryName);
+            }
+            else
+            {
+                return NullLogger.Instance;
+            }
+        }
+
+        public void ProcessBufferedLogs(IEnumerable<ILoggerProvider> forwardingProviders, bool runImmediately = false)
+        {
+            // Disable the channel and let the consumer know that there won't be any more logs.
+            _isEnabled = false;
+            _channel.Writer.TryComplete();
+
+            // Forward all buffered logs to the new provider
+            Task.Run(async () =>
+            {
+                if (!runImmediately)
+                {
+                    // Wait for 10 seconds, this will increase the probability of these logs appearing in live metrics.
+                    await Task.Delay(10000);
```

Comment:
nit: can we make this a readonly or const ? I.e. 

`private readonly TimeSpan DeferredLogDelay = TimeSpan.FromSeconds(10);`

...

`await Task.Delay(DeferredLogDelay)`

Reply:
Done.